### PR TITLE
Added OSD tag support

### PR DESCRIPTION
**Pull request by bond95**
# Description

With this feature is able to return only rules that could be fixed on OSD clusters.

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

## Testing steps

`make test`

### CODE REVIEW

File: content/content.go [hunk 1]
```diff
@@ -43,7 +44,7 @@ type ruleIDAndErrorKey struct {
 // It's thread safe
 type RulesWithContentStorage struct {
 	sync.RWMutex
-	rulesWithContent map[ruleIDAndErrorKey]*types.RuleWithContent
+	rulesWithContent map[ruleIDAndErrorKey]*local_types.RuleWithContent
```
Comment (Sergey1011010) [inline content/content.go:47]: why did you change this?

File: server/server.go [hunk 2]
```diff
@@ -458,7 +459,7 @@ func (server HTTPServer) readAggregatorRuleForClusterID(
 	return aggregatorResponse.Report, true
 }
 
-func (server HTTPServer) fetchRuleContent(rule types.RuleOnReport) (proxy_types.RuleWithContentResponse, bool) {
+func (server HTTPServer) fetchRuleContent(rule types.RuleOnReport, OSDEligible bool) (proxy_types.RuleWithContentResponse, bool, bool) {
```
Comment (Sergey1011010) [inline server/server.go:462]: I think it's better to document the return options since it's not very obvious what they are without reading the code

File: server/server.go [hunk 3]
```diff
@@ -517,16 +522,32 @@ func (server HTTPServer) fetchAggregatorReport(
...
 	if !successful {
 		return
 	}
 
 	var rules []proxy_types.RuleWithContentResponse
+	hasNoErrors := true
 
 	for _, aggregatorRule := range aggregatorResponse.Report {
-		rule, successful := server.fetchRuleContent(aggregatorRule)
+		rule, successful, isError := server.fetchRuleContent(aggregatorRule, server.getOSDFlag(request))
```
Comment (Sergey1011010) [inline server/server.go:548]: what's the difference between successful and isError?

File: server/server.go [hunk 4]
```diff
@@ -535,13 +556,16 @@ func (server HTTPServer) reportEndpoint(writer http.ResponseWriter, request *htt
 		rules = append(rules, rule)
 	}
 
-	if len(aggregatorResponse.Report) != 0 && len(rules) == 0 {
+	if len(aggregatorResponse.Report) != 0 && !hasNoErrors {
```
Comment (Sergey1011010) [inline server/server.go:559]: why?

File: server/server.go [hunk 5]
```diff
@@ -591,14 +615,19 @@ func (server HTTPServer) singleRuleEndpoint(writer http.ResponseWriter, request
 		return
 	}
 
-	rule, successful = server.fetchRuleContent(*aggregatorResponse)
+	rule, successful, _ = server.fetchRuleContent(*aggregatorResponse, server.getOSDFlag(request))
 
 	if !successful {
+		err := responses.SendOK(writer, responses.BuildResponse("Rule was not found"))
```
Comment (Sergey1011010) [inline server/server.go:621]: How can it be 200 when the rule wasn't found?

File: content/content.go (hunk 1 above)
Comment (bond95) [reply content/content.go:47]: Because I use local `RuleWithContent` which is used only in smart-proxy

File: server/server.go (hunk 2 above)
Comment (bond95) [reply server/server.go:462]: Yes, sure

File: server/server.go (hunk 3 above)
Comment (bond95) [reply server/server.go:548]: `successful` indicates if rule content was fetched (it could be filtered because of OSD flag). `isError` indicate if there appear some error, for example during parsing.

File: server/server.go (hunk 4 above)
Comment (bond95) [reply server/server.go:559]: Because `len(rules)` could be empty because of filtering, but also because of `errors`, so this `if` should be triggered only in case of errors.

File: server/server.go (hunk 5 above)
Comment (bond95) [reply server/server.go:621]: Oops, you're right

File: server/server.go (hunk 2 above)
Comment (tisnik) [reply server/server.go:467]: dunno if it is not better to return error value with proper information about the failure

File: server/server.go (hunk 3 above)
Comment (tisnik) [reply server/server.go:553]: I'd prefer to use `error` instead, WDYT?

File: server/server.go [hunk 6]
```diff
@@ -458,7 +459,12 @@ func (server HTTPServer) readAggregatorRuleForClusterID(
 	return aggregatorResponse.Report, true
 }
 
-func (server HTTPServer) fetchRuleContent(rule types.RuleOnReport) (proxy_types.RuleWithContentResponse, bool) {
+// fetchRuleContent - fetching content for particular rule
+// Return values:
+//   - Structure with rules and content
+//   - return true if fetching content was successful, including filtering
+//   - return true if fetching and parsing of content was successful, not including filtering
+func (server HTTPServer) fetchRuleContent(rule types.RuleOnReport, OSDEligible bool) (proxy_types.RuleWithContentResponse, bool, bool) {
```
Comment (tisnik) [inline server/server.go:467]: at least the last `bool` seems it should be in fact `error`

File: server/server.go (hunk 4 above)
Comment (Sergey1011010) [reply server/server.go:564]: but you could write an error earlier, when it happens in the loop and stop the whole thing, you ignore rules anyways

File: server/server.go [hunk 7]
```diff
@@ -489,7 +499,7 @@ func (server HTTPServer) fetchRuleContent(rule types.RuleOnReport) (proxy_types.
 		Internal:        ruleWithContent.Internal,
 	}
 
-	return parsedRule, true
+	return parsedRule, true, true
```
Comment (Sergey1011010) [inline server/server.go:502]: Why 3rd one is true? According to the code below it is "isError", but it doesn't look like any error happened here.

Comment (Sergey1011010) [reply server/server.go:502]: also, having 2 return values indicating errors by opposite values(successful is true = "ok", isError is false = "ok") is kinda confusing

File: server/server.go (hunk 4 above)
Comment (Sergey1011010) [reply server/server.go:564]: btw, do you really need to return an error if this condition `if OSDEligible && !ruleWithContent.NotRequireAdmin {` is met only for one rule? Maybe just ignore this rule

File: content/parsing.go [hunk 8]
```diff
@@ -31,6 +32,15 @@ var (
...
 
+func contains(arr []string, str string) bool {
+	for _, a := range arr {
+		if a == str {
+			return true
+		}
+	}
+	return false
+}
+
```
[the remark this answers is not in the document]
Comment (bond95) [reply content/parsing.go:43]: Yes, we have, thank you

File: server/server.go (hunk 6 above)
Comment (bond95) [reply server/server.go:467]: I'm not sure about that, because `fetchRuleContent` handling errors by itself, so in that case we just need to be sure if there is some errors.

File: server/server.go (hunk 4 above)
Comment (bond95) [reply server/server.go:564]: Yes, and no, because if we handle errors in this place, will mean, that in logs we get all rules, that's couldn't be parsed or fetched, which is much more better than stop request after first error.

Comment (bond95) [reply server/server.go:564]: It's not returning error when it's OSD, it just return not successful state, which mean that this rule will be ignored.

File: server/server.go (hunk 2 above)
Comment (bond95) [reply server/server.go:467]: But information about failure is handled before in inner functions.

File: server/server.go (hunk 7 above)
Comment (bond95) [reply server/server.go:502]: Fixed variable name.

File: server/server.go (hunk 2 above)
Comment (tisnik) [reply server/server.go:467]: ok agree. Could you then please name the return values? TYVM

File: content/parsing.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (tisnik) [reply content/parsing.go]: ok, thank you

File: server/server.go (hunk 2 above)
Comment (bond95) [reply server/server.go:467]: @tisnik I already left a comment, is there enough, check it please?

Comment (tisnik) [reply server/server.go:467]: ok not a problem

**Review by tisnik — APPROVED** [review]
LGTM

File: server/server.go (hunk 4 above)
Comment (Sergey1011010) [reply server/server.go:564]: Now you changed the logic and the error will be returned to the user if a single rule couldn't parse, not sure if we want this logic, but previous version was different(we just logged this error and returned what we could). @tisnik wdyt?

File: server/server.go [hunk 9]
```diff
@@ -591,14 +620,19 @@ func (server HTTPServer) singleRuleEndpoint(writer http.ResponseWriter, request
 		return
 	}
 
-	rule, successful = server.fetchRuleContent(*aggregatorResponse)
+	rule, successful, _ = server.fetchRuleContent(*aggregatorResponse, server.getOSDFlag(request))
```
Comment (Sergey1011010) [inline server/server.go:623]: Why do you ignore error here?

File: server/server.go [hunk 10]
```diff
@@ -591,14 +620,19 @@ func (server HTTPServer) singleRuleEndpoint(writer http.ResponseWriter, request
...
 
 	if !successful {
+		err := responses.SendNotFound(writer, "Rule was not found")
```
Comment (Sergey1011010) [inline server/server.go:626]: You could use only handleServerError here https://github.com/RedHatInsights/insights-operator-utils/blob/master/types/errors.go#L91

File: server/server.go (hunk 9 above)
Comment (bond95) [reply server/server.go:623]: Because if at least something goes wrong then `successful` will be `false`, including errors. So it's no need to handle it.

File: server/server.go (hunk 10 above)
Comment (bond95) [reply server/server.go:626]: `singleRuleEndpoint` don't get ruleID and ruleError, so using extracting that params second time just for this error, I think it's don't worth it https://github.com/RedHatInsights/insights-operator-utils/blob/15514d9c64e25edf80fee81e229b74c4b70b27e5/types/errors.go#L110

File: server/server.go (hunk 9 above)
Comment (Sergey1011010) [reply server/server.go:623]: well, if somebody changes the implementation, then there will be a problem

File: server/server.go (hunk 10 above)
Comment (Sergey1011010) [reply server/server.go:626]: `aggregatorResponse.Module`, but ok, not a big deal

**Review by joselsegura — APPROVED** [review]
LGTM

File: config.toml [hunk 11]
```diff
@@ -3,8 +3,8 @@ address = ":8081"
 api_prefix = "/api/v1/"
 api_spec_file = "openapi.json"
 debug = true
-auth = false
-auth_type = "xrh"
+auth = true
+auth_type = "xhr"
```
Comment (quarckster) [inline config.toml:7]: Why did you change it?

**Review by quarckster — CHANGES_REQUESTED** [review]

File: config.toml (hunk 11 above)
Comment (bond95) [reply config.toml:7]: I'll change it back. I used it for local testing

Comment (bond95) [reply config.toml]: Fixed

**Review by quarckster — APPROVED** [review]
